# ÜNİTE 6- Koşullu Durumlar ve Döngüler


## 1. if Bloğu

Yaşamın her anında ve bilgisayar programlarında, bir sonraki eylem her zaman "eğer" (İngilizce'de if) ile başlayan bir sorunun cevabına bağlıdır. Bu durum, eylemin sonuçlarının belirli bir kritere bağlı olarak çeşitlilik gösterebileceğini anlatır. 

Koşullu durumlara giriş yapmak adına, program kullanıcısının girdiği deniz suyu sıcaklığına bağlı olarak, denize girip girmememiz gerektiğini bize sunan küçük bir program yazalım.


In [1]:
T = float(input("Deniz suyu sıcaklığı nedir? "))
if T > 24:
    print("Harika! Hemen suya atla!")


Deniz suyu sıcaklığı nedir? 22


if bloğu koşulun sağlanıp sağlanamadığını araştırır. Eğer if sözcüğünden sonra belirtilen koşul sağlanırsa `(True)` if bloğu çalışır, koşul sağlanmazsa `(False)` if bloğu çalışmaz. Üstteki örnekte görüleceği üzere, Python öncelikle deniz suyu sıcaklığını kullanıcıya sorar. Örnekteki gibi bu değer 25 derece olarak girildiğinde, if sözcüğünden sonra belirtilen `T < 24 `koşulu sağlanacağı için if bloğu çalışır ve program bize hemen suya atlamamızı öneren bir metin gösterir. Şayet kullanıcı 24 dereceden daha küçük bir değer girseydi, koşul sağlanamayacağından if bloğu çalışmaz ve ekranda herhangi bir çıktı göremeyecektik. 

Bu tek koşula bağlı durum oldukça normaldir. Ancak koşul sağlanamadığında, Python’un bize bir çıktı üretmemesi programlama adına isteyebileceğimiz bir durum değildir. Bu yüzden, deniz suyu sıcaklığının 24 derecenin altında olduğu durumlar için de çıktı üretebilen bir programa gereksinmemiz bulunmaktadır. Programı yeniden düzenleyelim;


In [2]:
T = float(input("Deniz suyu sıcaklığı nedir? "))
if T > 24:
    print("Harika! Hemen suya atla!")
if T <= 24: 
    print("Suya girme. Çok soğuk!")


Deniz suyu sıcaklığı nedir? 22
Suya girme. Çok soğuk!


Bu örnekte de görüleceği üzere, Python ilk if bloğundan `False` sonucunu çıkardıktan sonra, ikinci if bloğundaki koşulu denetler. Bu program, ekrana istediğimiz sonuçları vermesine rağmen yine de eksik bir yön barındırmaktadır. İkinci sorgunun `True` çıkacağını önceden bildiğimiz için, bu iki koşullu if bloklarını kullanmak gereksizdir. 

## 2. if-else Bloğu

İki ayrı if bloğu oluşturmak yerine, deniz suyu sıcaklığı örneği için daha mantıklı bir programlama biçimi olan if-else bloğunu kullanalım. 

In [3]:
T = float(input("Deniz suyu sıcaklığı nedir? "))
if T > 24: # Koşulu test ediyoruz.
    print("Harika! Hemen suya atla!")
else:
    print("Suya girme. Çok soğuk!")

Deniz suyu sıcaklığı nedir? 22
Suya girme. Çok soğuk!


Bu programda `else` sözcüğünü kullanarak, ilk sıradaki  `if` ile başlayan koşulun `False` oluşması ile Python’un hemen `else` sözcüğünden sonra gelen çıktıyı ekrana vermesi sağlanır. Bu sayede, deniz suyu sıcaklığını ikinci bir koşula gerek duymadan denetlemiş oluruz. Buradaki mantık şudur: “Eğer ilk koşul yanlış ise (`False`), diğer koşul mutlaka doğru olmak zorundadır (`True`). 


## 3. if-elif-else Bloğu

Üstteki örneklerde deniz suyu sıcaklığı için ya soğuk ya da ılık olduğuna dair, 24 dereceyi eşik kabul eden tek bir koşul kullandık. Ancak gerçek hayatta her zaman tek koşuldan meydana gelen örnekler ile karşılaşmayız. İlk örnekte gerçekleştirdiğimiz 24 derecelik eşiği kullanmayalım. Sıradaki programımız 24 derece sıcaklığın yukarısındaki deniz suyunun yüzme için uygun olduğunu; 20-24 derece aralığındaki deniz suyu sıcaklığına “temkinli olunması” gerektiğini; 20 derece sıcaklığın altı için de “soğuk ve yüzülemez” olduğunu belirtsin. 


In [4]:
T = float(input("Deniz suyu sıcaklığı nedir? "))
if T > 24: # koşul 1
    print("Harika! Hemen suya atla!")
elif 20 <= T <= 24: # koşul 2
    print("Dikkat et! Önce ayağını sok")
else:
    print("Suya girme. Çok soğuk!")

Deniz suyu sıcaklığı nedir? 22
Dikkat et! Önce ayağını sok


Bu programda görüleceği üzere, if sözcüğünden sonra gelen koşul `False` olarak değerlendirildiği için program otomatik olarak `elif` sözcüğünden sonraki koşulu denetlemeye geçmiştir. `elif` sözcüğünde belirtilen koşul `True` olarak sağlandığı için, program `else` sözcüğüne geçmeden `elif` koşulunun sağlanması durumunda ekrana verilecek olan çıktı program tarafından gösterilmiştir.

Bir başka örnek de, öğrencinin başarı puanına göre harf notunun ne olacağına ilişkin sonucu üreten programı yazmakla verelim:

In [5]:
puan = 77
if puan >= 90:
    print('A')
elif puan >= 80:
    print('B')
elif puan >= 70:
    print('C')
elif puan >= 60:
    print('D')
else:
    print('F')

C


Bu örnekte de görüleceği üzere birden fazla `elif` koşulu oluşturulmuştur. Program `True` sonucunu buluncaya kadar çalışmaya devam edecektir. `True` sonucuna eriştiği yerde, program çalışmayı durdurup, o koşulun altında bulunan çıktıyı ekrana verecektir. Başarı puanı örneğinde, 77 alan bir öğrencinin harf notu da `C` olarak ekrana verilecektir. 

## 4. while Döngüsü

if-elif-else bloklarına girilen veri, bu koşul sorgulamasından sadece bir kez geçer. Yani program, girilen veriyi sadece bir kez denetimden geçirir. Ancak bazı durumlarda, girilen veri koşulu sağlayıncaya kadar denetimden geçirilmelidir. Bunun için while döngüsü kullanılır.
Örneğin, hata miktarı 50 olsun. Her döngüde hata kendini 4’e bölsün. Hata miktarı 1’den küçük oluncaya kadar döngü devam etsin.


In [6]:
hata = 50.0
while hata > 2 :
    hata = hata / 2
    print(hata)



25.0
12.5
6.25
3.125
1.5625


İlk turda 50 olan hata miktarı dörde bölünmüş ve ekrana 12.5 sonucu verilmiştir. Sonuç hala 1’den büyüktür. İkinci turda ise 12.5 olan hata miktarı yine 4’e bölünmüş ve 3.125 sonucu ekrana verilmiştir. Yine hata miktarı 1’den büyük olduğu için döngü devam edecektir. Üçüncü turda ise 3.125 olan hata miktarı 4’e bölünerek ekrana 0.78125 verilmiştir. Ancak elde edilen yeni hata miktarı 1’den küçük olduğu için while döngüsü sonlanmıştır. Yeni bir tur olmayacaktır. 
while döngüsü kullanılırken dikkat edilmesi gereken nokta, belirtilen koşulun (örnekte dörde bölme işlemi) mutlaka bir noktada son bulması gerektiğidir (Diğer bir deyişle, mutlaka bir noktada `False` sonucunu vermelidir). Aksi takdirde, while döngüsü sonsuza kadar çalışmayı gerektirir. Bu da Python derleyicisinde bağlantı problemi oluşturur. 


## 5. for Döngüsü

Belirli sayıda bir işlemi yapmak ya da bir işlemi bir dizinin elemanları üzerinde gerçekleştirmek için kullanılan döngüdür. For döngülerinin yazımında, bir değişken ve bu değişkenin ilk elemandan son elemana kadar işlem uyguladığı bir dizi tanımlanmak zorundadır. 
         
         for değişken in dizi:
             işlem 
             
Örneğin; bir listedeki tüm elemanların kübünü alabilen bir program oluşturalım. Bu programı yazmadan önce `range() `fonksiyonunu açıklamakta fayda vardır. Bu fonksiyon, içinde belirtilen iki sayı arasında (ikincisini dahil etmeden) bir sayı dizisi oluşturmamıza yarar. Kübünü alacağımız sayı dizisi (2,3,4,5) olsun. Bunu oluşturmak için `range (2,6)` şeklinde bir fonksiyon tanımlamamız gerekir. Şayet, `range()` fonksiyonu içerisinde tek bir sayı varsa, bunun sonucu olarak sıfırdan başlayıp bu sayının bir eksiğinde sona eren bir sayı dizisi oluşur. Örneğimize geri dönecek olursak;

In [7]:
for i in range(2,6):
    print(i**3)

8
27
64
125


Bu döngüde `i` değişkeni, `range(2,6)` ise değişkenin üzerinde işlem yapacağı diziyi ifade etmektedir. Bu döngüde yapılacak işlem ise, dizideki her elemanın (i değişkeninin) kübünü alıp ekrana verme (print) işlemidir. 

Listeler üzerinde de döngüler oluşturmak olanaklıdır. Örneğin, bir ailenin tüm fertlerinin boylarının sıralandığı bir listenin tüm elemanlarını for döngüsü kullanarak ekrana yazdıralım. 

In [8]:
aile = [1.73, 1.68, 1.71, 1.89]
for boy in aile:
    print(boy)

1.73
1.68
1.71
1.89


Her zaman listelerimiz bu kadar küçük olmayabilir. Büyük veri çağında, birbirine ardışık olarak sıralanmış veri listelerini kullanmamız daha olasıdır. Dolayısıyla, liste elemanlarını for döngüsü ile ekrana yazdırırken, bu elemanların her birinin sırasını da (indeks numarasını da) `enumerate()` fonksiyonu kullanarak ekrana vermemiz gerekebilir. Bu fonksiyon, hem listelerde, hem de sözlük ve dizilerde kullanılabilir. Bu sıralı veri yapılarındaki her bir sıraya "indeks" adını verelim. Üstteki aile fertlerinin boyunu içeren listeyi kullanarak, hem sıra numaralarını hem de boyları ekrana verelim. 

In [9]:
aile = [1.73, 1.68, 1.71, 1.89]
for indeks, boy in enumerate(aile):
    print("kişi " + str(indeks) + ": " + str(boy))

kişi 0: 1.73
kişi 1: 1.68
kişi 2: 1.71
kişi 3: 1.89


Görüldüğü üzere hem `indeks`, hem `boy` değişkenleri `enumerate()` fonksiyonu ile numaralandırdığımız `aile` listesi üzerinde işlem yapsın diye for döngüsünü oluşturduk. Döngü işlem olarak `"kişi " + str(indeks) + ": " + str(boy)` şeklinde tüm liste elemanlarını sırayla ekrana vermektedir. Ancak `"kişi 0"` kullanımı biraz tuhaf bir görünüme sebep vermiştir. Ekrana verme işlemini `"kişi 1"` ile başlayacak şekilde gerçekleştirmek için aşağıdaki düzenlemeyi yapalım. 

In [10]:
aile = [1.73, 1.68, 1.71, 1.89]
for indeks, boy in enumerate(aile):
    print("kişi " + str(indeks + 1) + ": " + str(boy))

kişi 1: 1.73
kişi 2: 1.68
kişi 3: 1.71
kişi 4: 1.89


Listelerden oluşan listeler için de for döngüsü kullanılabilmektedir. En kalabalık beş kentimizi ve nüfuslarını içeren listelerden oluşan listeyi kullanalım: 

In [11]:
nufuslar = [['İstanbul', 15519267],
           ['Ankara', 5639076],
           ['İzmir', 4367251],
           ['Bursa', 3056120],
           ['Antalya', 2511700]]

for i in nufuslar: 
    print(str(i[0] + ": " + str(i[1])))

İstanbul: 15519267
Ankara: 5639076
İzmir: 4367251
Bursa: 3056120
Antalya: 2511700


For döngüsünün değişkeni olan `i`, bu listelerden oluşan listenin içindeki her bir listedir. `i[0]` ise her bir listenin içindeki ilk elemandır. Bu örnekte, kent ismini belirten metinlerdir. `i[1]` ise, listelerin ikinci elemanı olup; bu örnekte, kentin nüfusunu belirten tam sayılardır. 



## 6. Yineleyiciler (İteratör)

for döngüsü anlatılırken sırayla ve tek tek işlem yapılan nesnelerden bahsedilmişti. Bu nesnelerin, yinelenen (iterable) bir yapıya sahip olduğu söylenebilir. Bir başka deyişle, burada yinelenen bir nesne üzerinde yapılan işlemdir. Yinelenen işlemi gerçekleştiren nesneye de yineleyici (iteratör) ismi verilir. Python'da yineleyici oluşturmak için `iter()` fonksiyonu, yinelenen nesnedeki elemanlara erişmek için de `next()` fonksiyonu kullanılmaktadır. 

![figure](https://i.hizliresim.com/Rs2H1n.png)

Oluşturulan listedeki her bir elemana (metin nesnesine) ayrı ayrı ulaşmayı hedefliyoruz. `iter()` fonksiyonunu kullanarak bir yineleyici (iterator) oluşturuldu. Ardından `next()` fonksiyonu kullanarak her bir elemana erişilir. Bu elemanlar farklı farklı veri yapılarında olabilir. Bu örnekteki tüm veri yapıları metin tipidir. `next()` fonksiyonu ile 't' elemanına eriştikten sonra, tekrar `next()` fonksiyonu kullanılırsa, `StopIteration` hatası görülür. Çünkü son elemandan sonra, artık `next()` fonksiyonu ile erişilecek başka eleman kalmamıştır. 

Yinelenmek istenen nesnedeki tüm elemanlara aynı anda erişebilmek için yineleyici önüne `*` işareti eklenmelidir. Örneğin, 

In [12]:
isim = "Python"
yineleyici = iter(isim)
print(*yineleyici)

P y t h o n


Programlama mantığında bir nesne üzerinde tüm elemanlarına erişecek şekilde yineleme işlemini uygulamak yeterli gelmeyecektir. Genellikle belirli sayıda işlemi arka arkaya yaptırmak daha çok tercih edilir. Örneğin range() fonksiyonu kullanarak sayı listesindeki elemanları yineleyen bir yineleyici nesne oluşturalım. `range(x)` kullanımı, 0'dan başlayıp x'in bir eksiğine (x-1) kadar bir sayı listesi oluştururmuş gibi çalışır. 

In [13]:
for i in range(4): 
    print(i)

0
1
2
3


`range()` nesnesi `print()` fonksiyonu ile ekrana yazdırılmaya çalışıldığında, herhangi bir sayı listesi sonucuna ulaşmak olanaklı değildir. 

In [14]:
aralik = range(5,7)
print(aralik)

range(5, 7)


`range()` fonksiyonu ile oluşturulan nesnenin içerdiği sayıları ekrana verebilmek için, bu nesneyi `list()` ya da `set()` fonksiyonları kullanarak liste ya da küme veri yapısına çevirmek gerekmektedir. 

In [15]:
aralik = range(5,7)
aralik_listesi = list(aralik)
print(aralik_listesi)

[5, 6]


Yineleyiciler konusunda kullanılan bir diğer önemli fonksiyon da `zip()` fonksiyonudur. Bu fonksiyon, bir ya da daha fazla yinelenen nesneyi (iterable) argüman olarak alır ve demet (tuple) dizileri üretir. Örneğin; 

In [16]:
kentler = ['İstanbul','Ankara', 'İzmir','Bursa', 'Antalya']
nufus = ['15519267','5639076', '4367251', '3056120', '2511700']

for kent, nufusu in zip(kentler,nufus):
    print(kent,nufusu)

İstanbul 15519267
Ankara 5639076
İzmir 4367251
Bursa 3056120
Antalya 2511700


`zip()` fonksiyonu ile "fermuarlanmış" diziler, kendilerini oluşturan bileşenlere `*` sembolü kullanılarak ayrılabilir. 

In [17]:
kentler = ['İstanbul','Ankara', 'İzmir','Bursa', 'Antalya']
nufus = ['15519267','5639076', '4367251', '3056120', '2511700']

kent_nufusu = zip(kentler, nufus)

sehir, insan_sayisi = zip(* kent_nufusu)
print(sehir)
print(insan_sayisi)

('İstanbul', 'Ankara', 'İzmir', 'Bursa', 'Antalya')
('15519267', '5639076', '4367251', '3056120', '2511700')


## 7. List Comprehension

for döngüsü kullanılarak yeni bir liste oluşturmaya çalışalım. for döngüsünün oluşturduğu her bir eleman `.append()` yöntemi ile oluşturduğumuz yeni listeye eklensin. Döngü tamamlandığında da oluşturduğumuz liste nihai şekliyle ekrana verilsin. 

In [18]:
# Boş bir liste oluşturalım

kup = []

# for döngüsünde kullandığımız küp almamızı sağlayan döngüyü burada da kullanalım.

for i in range(2,6):
    kup.append(i**3)
print(kup)

[8, 27, 64, 125]


Bu denli basit bir işlemi, bu kadar fazla satırda kod kullanarak yapmak zorlayıcı ve yavaş bir tercih olacaktır. Bunun yerine list comprehension (liste döngüleri) kullanmak, aynı işlemi tek satır kod yazarak gerçekleştirmemizi sağlayacaktır. 

In [19]:
kup = [i ** 3 for i in range(2,6)]
print(kup)

[8, 27, 64, 125]


Görüleceği üzere, list comprehension kullanımı, iki köşeli parantez arasında bir for döngüsü oluşturmamızı sağlamaktadır. 

    [çıktı ifade for yineleyici in yinelenen nesne]
    
Bir örnek daha yapalım: 

In [20]:
liste1 = [1,2,3]
liste2 = [4,5,6]
combo = [(x,y) for x in liste1 for y in liste2]
print(combo)

[(1, 4), (1, 5), (1, 6), (2, 4), (2, 5), (2, 6), (3, 4), (3, 5), (3, 6)]


Bu örnekte ise iki liste tanımladık. İlk değişkeni x ve ikinci değişkeni y olarak tanımladık. x değişkeni liste1 elemanları, y değişkeni liste2 elemanları olacak şekilde ilk listenin elemanlarını ikinci listenin elemanları ile bir araya getiren dizilerin listesini oluşturduk. 

List comprehension kullanılırken koşul ifadeleri de kullanılabilir. 

    [çıktı ifadesi for yineleyici değişken in yinelenen nesne if koşul]
    
Beş kent isminden oluşan listedeki "İ" harfi ile başlayan kentlerin ilk üç harfini ekrana verelim.
   

In [21]:
kentler = ['İstanbul','Ankara', 'İzmir','Bursa', 'Antalya']

kentler_i = [kent[0:3] for kent in kentler if kent[0] == "İ"]

print (kentler_i)

['İst', 'İzm']


List comprehension içerisinde if ifadesi kullanılabildiği gibi, else ifadesi de kullanılabilir. Örneğin, 0'dan 7'ye kadar olan sayıların üzerinde işlem yaparak, bu sayıların hangisinin çift sayı hangisinin tek sayı olduğunu ekrana verelim. 

In [22]:
["Çift" if i%2 == 0 else "Tek" for i in range(8)]

['Çift', 'Tek', 'Çift', 'Tek', 'Çift', 'Tek', 'Çift', 'Tek']

## 8. İç İçe for Döngüleri

İç içe for döngüleri de kullanılabilmektedir. Bu iç içelik bir döngü içinde başka bir döngünün de kullanıldığını işaret eder. İçteki döngü, dıştaki döngünün her bir yinelemesi için tek bir sefer kullanılacaktır. Örneğin aşağıdaki listeleri kullanarak, her bir sıfatı tüm meyveler için ekrana yazdıralım.

In [23]:
sifat = ["kirmizi", "büyük", "lezzetli"]
meyve = ["elma", "muz", "vişne"]

for x in sifat:
    for y in meyve:
        print(x,y)

kirmizi elma
kirmizi muz
kirmizi vişne
büyük elma
büyük muz
büyük vişne
lezzetli elma
lezzetli muz
lezzetli vişne


Başka bir örnek verelim. Çarpım tablosunun 6'lı ve 7'li çarpım değerlerini ekrana yazdıralım. 

In [24]:
for i in range (6,8):
    for j in range(1,11):
        print(f"{i}*{j} = {i*j}")

6*1 = 6
6*2 = 12
6*3 = 18
6*4 = 24
6*5 = 30
6*6 = 36
6*7 = 42
6*8 = 48
6*9 = 54
6*10 = 60
7*1 = 7
7*2 = 14
7*3 = 21
7*4 = 28
7*5 = 35
7*6 = 42
7*7 = 49
7*8 = 56
7*9 = 63
7*10 = 70


## 9. Python'da Döngü Kontrolleri

Bazen normal akışındaki döngüye (iterasyona) müdahale etmemiz gerekebilir. Döngüye müdahale edebileceğimiz iki ana komut bulunmaktadır: `break` ve `continue`

break komutunun kullanıldığı koşul, döngü içerisinde gerçekleştiğinde döngü duracaktır. Örneğin:

In [25]:
for i in "Mersin":
    print(i)
    if i == "r": break;

M
e
r


continue komutunun birlikte kullanıldığı koşul döngü içerisinde sağlandığında, döngüdeki işlem o noktada çalışmaz ve döngü diğer elemanlara atlayarak çalışmaya devam eder. Örneğin: 

In [26]:
i = 0
while (i<8):
    i+=1
    if (i == 6): continue
    print(i)

1
2
3
4
5
7
8


Örnekte de görüleceği üzere, döngü değişkeni (i) 6'ya eşit olduğunda döngü bu noktada çalışmaz ve diğer elemanla çalışmaya devam eder. 